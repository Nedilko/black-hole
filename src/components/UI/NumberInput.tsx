import React, { memo, useCallback, useEffect, useRef } from 'react';
import { MdKeyboardArrowDown, MdKeyboardArrowUp } from 'react-icons/md';

type PropsType = {
  label: string;
  min: number;
  max: number;
  value: number;
  handleChange: (value: number) => void;
};

const NumberInput = ({ label, min, max, value, handleChange }: PropsType) => {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    const ignoreScroll = (e: WheelEvent) => {
      e.preventDefault();
    };
    inputRef.current &&
      inputRef.current.addEventListener('wheel', ignoreScroll);
  }, [inputRef]);

  const changeValueHandler = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      handleChange(+e.target.value);
    },
    [handleChange]
  );

  const handleIncrement = useCallback(() => {
    handleChange(value + 1);
  }, [handleChange, value]);

  const handleDecrement = useCallback(() => {
    handleChange(value - 1);
  }, [handleChange, value]);

  return (
    <label className="flex text-gray-300 uppercase relative w-full">
      <MdKeyboardArrowUp
        data-testid="increment"
        onClick={handleIncrement}
        className="absolute top-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
      />
      <MdKeyboardArrowDown
        data-testid="decrement"
        onClick={handleDecrement}
        className="absolute bottom-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
      />
      <div className="text-sm absolute top-1 left-2 text-gray-400 select-none">
        {label}
      </div>
      <input
        className="appearance-none outline outline-1 hover:outline-2 outline-slate-700 focus:outline-gray-400 outline-offset-2 w-full text-xl font-semibold bg-slate-900 px-2 pt-6 pb-1 rounded transition-all duration-150"
        type="number"
        min={min}
        max={max}
        onChange={changeValueHandler}
        value={value}
        ref={inputRef}
      />
    </label>
  );
};

export default memo(NumberInput);
