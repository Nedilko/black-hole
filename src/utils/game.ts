import { DIFFUCULTY, Results } from '../store/gameSlice';

export const getDefaultResults = (): Results => {
  return {
    [DIFFUCULTY.EASY]: 0,
    [DIFFUCULTY.MEDIUM]: 0,
    [DIFFUCULTY.HARD]: 0,
    [DIFFUCULTY.CUSTOM]: 0,
  };
};
