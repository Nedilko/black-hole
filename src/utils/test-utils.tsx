import { RenderOptions } from '@testing-library/react';
import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { PreloadedState } from 'redux';
import { AppStore, RootState, setupStore } from '../store';
import { render } from '@testing-library/react';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = setupStore(initialState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
};
