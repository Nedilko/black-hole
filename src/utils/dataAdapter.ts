import type { Results } from '../store/gameSlice';
import { getDefaultResults } from './game';
import {
  writeToLocalstorage,
  readFromLocalstorage,
} from './localstorageAdapter';

const loadResults = (): Results => {
  const localStorageData = readFromLocalstorage('results');
  if (!localStorageData) {
    writeResults(getDefaultResults());
    return getDefaultResults();
  }
  return JSON.parse(atob(localStorageData));
};

const writeResults = (settings: Results): void => {
  writeToLocalstorage('results', btoa(JSON.stringify(settings)));
};

export { loadResults, writeResults };
