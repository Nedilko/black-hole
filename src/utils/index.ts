import { getTime } from './time';
import {
  readFromLocalstorage,
  writeToLocalstorage,
} from './localstorageAdapter';
import { getDefaultResults } from './game';
import { loadResults, writeResults } from './dataAdapter';

export {
  getTime,
  readFromLocalstorage,
  writeToLocalstorage,
  getDefaultResults,
  loadResults,
  writeResults,
};
