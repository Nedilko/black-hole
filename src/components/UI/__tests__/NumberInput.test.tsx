import { fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { renderWithProviders } from '../../../utils/test-utils';
import NumberInput from '../NumberInput';

describe('NumberInput', () => {
  const changeHandler = jest.fn();

  beforeEach(() => {
    changeHandler.mockClear();
  });

  it('should render NumberInput', () => {
    const { container } = renderWithProviders(
      <NumberInput
        label={''}
        min={0}
        max={0}
        value={0}
        handleChange={changeHandler}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should handle change', () => {
    renderWithProviders(
      <NumberInput
        label={'sample'}
        min={0}
        max={20}
        value={5}
        handleChange={changeHandler}
      />
    );

    fireEvent.change(screen.getByLabelText(/sample/i), {
      target: { value: '6' },
    });
    fireEvent.change(screen.getByLabelText(/sample/i), {
      target: { value: '7' },
    });
    expect(changeHandler).toHaveBeenCalledTimes(2);
  });

  it('should handle click increment button', () => {
    renderWithProviders(
      <NumberInput
        label={'sample'}
        min={0}
        max={20}
        value={5}
        handleChange={changeHandler}
      />
    );

    userEvent.click(screen.getByTestId(/increment/i));
    expect(changeHandler).toHaveBeenCalledTimes(1);
    expect(changeHandler).toHaveBeenCalledWith(6);
  });

  it('should handle click decrement button', () => {
    renderWithProviders(
      <NumberInput
        label={'sample'}
        min={0}
        max={20}
        value={5}
        handleChange={changeHandler}
      />
    );

    userEvent.click(screen.getByTestId(/decrement/i));
    expect(changeHandler).toHaveBeenCalledTimes(1);
    expect(changeHandler).toHaveBeenCalledWith(4);
  });

  it('should prevent scrolling', () => {
    renderWithProviders(
      <NumberInput
        label={'sample'}
        min={0}
        max={20}
        value={5}
        handleChange={changeHandler}
      />
    );

    fireEvent.wheel(screen.getByLabelText(/sample/i), {
      deltaY: -1,
    });
    expect(changeHandler).not.toHaveBeenCalled();
  });
});
