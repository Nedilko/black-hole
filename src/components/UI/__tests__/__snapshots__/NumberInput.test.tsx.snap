// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput should render NumberInput 1`] = `
<div>
  <label
    class="flex text-gray-300 uppercase relative w-full"
  >
    <svg
      class="absolute top-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
      data-testid="increment"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
      />
    </svg>
    <svg
      class="absolute bottom-0 right-0 cursor-pointer text-3xl hover:scale-110 transition-all duration-150"
      data-testid="decrement"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
      />
    </svg>
    <div
      class="text-sm absolute top-1 left-2 text-gray-400 select-none"
    />
    <input
      class="appearance-none outline outline-1 hover:outline-2 outline-slate-700 focus:outline-gray-400 outline-offset-2 w-full text-xl font-semibold bg-slate-900 px-2 pt-6 pb-1 rounded transition-all duration-150"
      max="0"
      min="0"
      type="number"
      value="0"
    />
  </label>
</div>
`;
